pallets/dca/src/lib.rs

```diff
@@ -0,0 +1,1029 @@
+
+			Self::deposit_event(Event::Scheduled {
+				id: next_schedule_id,
+				who,
+			});
+
+			Ok(())
+		}
+
+		/// Admin endpoint to terminate a DCA schedule and remove it completely from the chain.
+		///
+		/// Parameters:
+		/// - `origin`: schedule owner
+		/// - `schedule_id`: schedule id
+		/// - `next_execution_block`: block number where the schedule is planned.
+		///
+		/// Emits `Terminated` event when successful.
+		#[pallet::call_index(1)]
+		#[pallet::weight(<T as Config>::WeightInfo::terminate())]
+		#[transactional]
+		pub fn terminate(
+			origin: OriginFor<T>,
+			schedule_id: ScheduleId,
+			next_execution_block: BlockNumberFor<T>,
+		) -> DispatchResult {
+			let ensure_technical_origin = T::TechnicalOrigin::ensure_origin(origin.clone());
+			let ensure_signed = ensure_signed(origin);
+			if ensure_technical_origin.is_err() && ensure_signed.is_err() {
+				return Err(BadOrigin);
+			}
+
+			let schedule = Schedules::<T>::get(schedule_id).ok_or(Error::<T>::ScheduleNotFound)?;
+
+			if let Ok(who) = ensure_signed {
+				ensure!(who == schedule.owner, Error::<T>::Forbidden);
+			}
+
+			Self::try_unreserve_all(schedule_id, &schedule);
+
+			let schedule_ids_on_block = ScheduleIdsPerBlock::<T>::get(next_execution_block);
+			ensure!(
+				schedule_ids_on_block.contains(&schedule_id),
+				Error::<T>::ScheduleNotFound,
+			);
+
+			//Remove schedule id from next execution block
+			ScheduleIdsPerBlock::<T>::try_mutate_exists(
+				next_execution_block,
+				|maybe_schedule_ids| -> DispatchResult {
+					let schedule_ids = maybe_schedule_ids.as_mut().ok_or(Error::<T>::ScheduleNotFound)?;
+
+					let index = schedule_ids
+						.iter()
+						.position(|x| *x == schedule_id)
+						.ok_or(Error::<T>::ScheduleNotFound)?;
+
+					schedule_ids.remove(index);
+
+					if schedule_ids.is_empty() {
+						*maybe_schedule_ids = None;
+					}
+					Ok(())
+				},
+			)?;
+
+			Self::remove_schedule_from_storages(&schedule.owner, schedule_id);
+
+			Self::deposit_event(Event::Terminated {
+				id: schedule_id,
+				who: schedule.owner,
+				error: Error::<T>::ManuallyTerminated.into(),
+			});
+
+			Ok(())
+		}
+	}
+}
+
+impl<T: Config> Pallet<T>
+where
+	<T as pallet_route_executor::Config>::AssetId: From<<T as pallet::Config>::Asset>,
+	<T as pallet_route_executor::Config>::Balance: From<u128>,
+	u128: From<<T as pallet_route_executor::Config>::Balance>,
+{
+	fn prepare_schedule(
+		current_blocknumber: T::BlockNumber,
+		weight: &mut Weight,
+		schedule_id: ScheduleId,
+		schedule: &Schedule<T::AccountId, T::Asset, T::BlockNumber>,
+	) -> Result<T::BlockNumber, DispatchError> {
+		let weight_for_single_execution = <T as Config>::WeightInfo::on_initialize_with_one_trade();
+
+		weight.saturating_accrue(weight_for_single_execution);
+
+		Self::take_transaction_fee_from_user(schedule_id, schedule, weight_for_single_execution)?;
+
+		let next_execution_block = current_blocknumber
+			.checked_add(&schedule.period)
```

Comment:
changed it, and do the retry by using 10 blocks (short oracle)

Reply:
would be better to increase retry delay with retry count (same as you would do for API call for example)

Reply:
How about this:

Base retry delay is 10.

then we increase it with 10 * (2 ^ number of retry).

So in case of 3 retries (configured in hydra), we will do:

10
20
40

Wdyt?

Reply:
i agree